### DISCUSSION TITLE
How to get access to for-generated widgets?

### DISCUSSION THREAD

**Discussion by quendimax** [slint-ui/slint#2503] [Q&A]
Hello everyone!

I can't find a way how to get access to the widgets generated by means of _for_-statement:

    HorizontalLayout {
         for data in data-model: widget := MyWidget {
             // data initialisation
         }
    }

In my case I want to add key event handler to change focus between the created widgets. For example:

    FocusScope {
        HorizontalLayout { ... } // the code above

        key-pressed(event) => {
            if (event == SomeShortcut) {
                 items[the-next-widget-index].focus();
            }
        }
    }

Is it possible to get the `items` here? Or maybe there is a workaround?

And another question. What can I do with the `widget` identifier from the first snippet? Firstly I had thought, it was what I was looking for. But then I found that I can use it only inside the `MyWidget` definition. Does it make sense?

Thank you!

**Comment by tronical** [comment]
Hi!:)

It's not possible to index items. What should happen with an out of bounds access? We could panic maybe. But I'm not sure that's intuitive.

What we've typically done so far is to create two-way bindings to "communicate" from inside to the outside.

The widget name is accessible within the repeated element.

**Reply by quendimax** [reply]
> What should happen with an out of bounds access?

Agree. If we can, we should avoid this.

> What we've typically done so far is to create two-way bindings to "communicate" from inside to the outside.

But my case requires "communicate" from outside to the inside. And with typical named widget, it is easy. But with repeated widgets, I don't understand how to implement this syntactically, because the name "is accessible within the repeated element" only.

I tried to write something like this:

    component Parent inherits Rectangle
    {
        in property<[int]> data;
        callback focus-changed(int);
    
        for data[id] in data: widget := Rectangle {
            property<int> id: id;
            callback focus-changed <=> root.focus-changed;
            focus-changed(id) => /*handler*/ {
                if (id == self.id) {
                    self.focus();
                }
            }
        }
    }

but it is not parsed. So the question is how to implement the `handler` if I've already have two-way binding with `focus-changed` callback?
Another attempt to implement this is:

    component Parent inherits Rectangle
    {
        property<[int]> data;
        callback focus-changed(int);

        for data[id] in data: widget := Rectangle {
            property<int> id: id;
            root.focus-changed(id) => {
                if (id == self.id) {
                    self.focus();
                }
            }
        }
    }

but again it is incorrect. Any ideas?

**Reply by tronical** [reply]
Hm, I don't think that it's possible to invoke a callback that's "inside" from the outside. Perhaps you can tunnel the information you want to convey through the data model? @ogoffart do you have any idea?

**Comment by quendimax** [comment]
I want to reformulate my question. Is it possible to bind a callback from outer scope with a signal/function from an inner *unnamed* item? For example:

     component Outer {
        callback done;
    
        Rectangle {
            function do() {
                debug("Rectangle.do");
            }
        }
    }

How can I bind the `done` signal with the `do` function?
